---

title: predict


keywords: fastai
sidebar: home_sidebar

summary: "Wrap all components of making rule set 3 predictions into one function"
description: "Wrap all components of making rule set 3 predictions into one function"
nb_path: "04_predict.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 04_predict.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">rs3.targetdata</span> <span class="kn">import</span> <span class="n">write_transcript_data</span><span class="p">,</span> <span class="n">write_conservation_data</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">design_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s1">&#39;test_data/sgrna-designs.txt&#39;</span><span class="p">)</span>
<span class="n">sanger_activity</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;test_data/Behan2019_activity.csv&#39;</span><span class="p">)</span>
<span class="n">gecko_activity</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;test_data/Aguirre2016_activity.csv&#39;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">multiprocessing</span>
<span class="n">max_n_jobs</span> <span class="o">=</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="predict_seq_tracr" class="doc_header"><code>predict_seq_tracr</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/predict.py#L22" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>predict_seq_tracr</code>(<strong><code>design_df</code></strong>, <strong><code>tracr</code></strong>, <strong><code>context_col</code></strong>, <strong><code>ref_tracrs</code></strong>, <strong><code>n_jobs</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="combine_target_seq_scores" class="doc_header"><code>combine_target_seq_scores</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/predict.py#L28" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>combine_target_seq_scores</code>(<strong><code>design_df</code></strong>, <strong><code>tracr</code></strong>, <strong><code>target_score_col</code></strong>, <strong><code>lite</code></strong>)</p>
</blockquote>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="predict" class="doc_header"><code>predict</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/predict.py#L36" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>predict</code>(<strong><code>design_df</code></strong>, <strong><code>tracr</code></strong>=<em><code>None</code></em>, <strong><code>target</code></strong>=<em><code>False</code></em>, <strong><code>aa_seq_file</code></strong>=<em><code>None</code></em>, <strong><code>domain_file</code></strong>=<em><code>None</code></em>, <strong><code>conservatin_file</code></strong>=<em><code>None</code></em>, <strong><code>id_cols</code></strong>=<em><code>None</code></em>, <strong><code>context_col</code></strong>=<em><code>'sgRNA Context Sequence'</code></em>, <strong><code>transcript_id_col</code></strong>=<em><code>'Target Transcript'</code></em>, <strong><code>transcript_base_col</code></strong>=<em><code>'Transcript Base'</code></em>, <strong><code>transcript_len_col</code></strong>=<em><code>'Target Total Length'</code></em>, <strong><code>n_jobs_min</code></strong>=<em><code>1</code></em>, <strong><code>n_jobs_max</code></strong>=<em><code>1</code></em>, <strong><code>lite</code></strong>=<em><code>True</code></em>)</p>
</blockquote>
<p>Make predictions using RS3</p>
<p>:param design_df: DataFrame
:param tracr: str or list
:param target: bool, whether to include target scores
:param aa_seq_file: str, path to precomputed amino acid sequences
:param domain_file: str, path to precomputed domain file
:param id_cols: list or None
:param context_col: str
:param transcript_id_col: str
:param transcript_base_col: str
:param transcript_len_col: str
:param n_jobs_min: int
:return: DataFram</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">scored_designs</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="n">design_df</span><span class="p">,</span> <span class="n">tracr</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Hsu2013&#39;</span><span class="p">,</span> <span class="s1">&#39;Chen2013&#39;</span><span class="p">],</span> <span class="n">target</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                         <span class="n">n_jobs_min</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">n_jobs_max</span><span class="o">=</span><span class="n">max_n_jobs</span><span class="p">,</span>
                         <span class="n">lite</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Calculating sequence-based features
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 400/400 [00:03&lt;00:00, 113.01it/s]
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Calculating sequence-based features
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 400/400 [00:00&lt;00:00, 1934.24it/s]
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting amino acid sequences
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 4/4 [00:03&lt;00:00,  1.00it/s]
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting protein domains
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 200/200 [01:10&lt;00:00,  2.82it/s]
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting conservation
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 200/200 [03:47&lt;00:00,  1.14s/it]
/Users/pdeweird/Documents/rs3/rs3/targetdata.py:345: UserWarning: Failed to get conservation scores for 2 transcriptsENST00000368892, ENST00000629399
  warnings.warn(&#39;Failed to get conservation scores for &#39; + str(len(failed_list)) +
&lt;ipython-input-5-c4cb19900fff&gt;:92: UserWarning: Missing conservation scores for transcripts: ENST00000368892,ENST00000629399
  warnings.warn(&#39;Missing conservation scores for transcripts: &#39; +
/Users/pdeweird/opt/anaconda3/envs/rs3/lib/python3.8/site-packages/sklearn/base.py:310: UserWarning: Trying to unpickle estimator SimpleImputer from version 1.0.dev0 when using version 0.24.2. This might lead to breaking code or invalid results. Use at your own risk.
  warnings.warn(
/Users/pdeweird/opt/anaconda3/envs/rs3/lib/python3.8/site-packages/sklearn/base.py:310: UserWarning: Trying to unpickle estimator Pipeline from version 1.0.dev0 when using version 0.24.2. This might lead to breaking code or invalid results. Use at your own risk.
  warnings.warn(
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">write_transcript_data</span><span class="p">(</span><span class="n">design_df</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
                      <span class="n">filepath</span><span class="o">=</span><span class="s1">&#39;./data/target_data/&#39;</span><span class="p">,</span>
                      <span class="n">aa_seq_name</span><span class="o">=</span><span class="s1">&#39;aa_seqs.pq&#39;</span><span class="p">,</span>
                      <span class="n">protein_domain_name</span><span class="o">=</span><span class="s1">&#39;protein_domains.pq&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting amino acid sequences
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 4/4 [00:03&lt;00:00,  1.23it/s]
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting protein domains
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 200/200 [00:53&lt;00:00,  3.70it/s]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">write_conservation_data</span><span class="p">(</span><span class="n">design_df</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="n">max_n_jobs</span><span class="p">,</span>
                        <span class="n">filepath</span><span class="o">=</span><span class="s1">&#39;./data/target_data/&#39;</span><span class="p">,</span>
                        <span class="n">cons_file_name</span><span class="o">=</span><span class="s1">&#39;conservation.pq&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting conservation
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 200/200 [04:05&lt;00:00,  1.23s/it]
/Users/pdeweird/Documents/rs3/rs3/targetdata.py:345: UserWarning: Failed to get conservation scores for 1 transcriptsENST00000393047
  warnings.warn(&#39;Failed to get conservation scores for &#39; + str(len(failed_list)) +
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">scored_designs_stored</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="n">design_df</span><span class="p">,</span> <span class="n">tracr</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Hsu2013&#39;</span><span class="p">,</span> <span class="s1">&#39;Chen2013&#39;</span><span class="p">],</span> <span class="n">target</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                <span class="n">n_jobs_min</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">n_jobs_max</span><span class="o">=</span><span class="n">max_n_jobs</span><span class="p">,</span>
                                <span class="n">aa_seq_file</span><span class="o">=</span><span class="s1">&#39;./data/target_data/aa_seqs.pq&#39;</span><span class="p">,</span>
                                <span class="n">domain_file</span><span class="o">=</span><span class="s1">&#39;./data/target_data/protein_domains.pq&#39;</span><span class="p">,</span>
                                <span class="n">conservatin_file</span><span class="o">=</span><span class="s1">&#39;./data/target_data/conservation.pq&#39;</span><span class="p">,</span>
                                <span class="n">lite</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">pd</span><span class="o">.</span><span class="n">testing</span><span class="o">.</span><span class="n">assert_frame_equal</span><span class="p">(</span><span class="n">scored_designs</span><span class="p">,</span> <span class="n">scored_designs_stored</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Calculating sequence-based features
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 400/400 [00:00&lt;00:00, 605.59it/s]
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Calculating sequence-based features
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 400/400 [00:00&lt;00:00, 2427.89it/s]
&lt;ipython-input-5-c4cb19900fff&gt;:92: UserWarning: Missing conservation scores for transcripts: ENST00000393047
  warnings.warn(&#39;Missing conservation scores for transcripts: &#39; +
/Users/pdeweird/opt/anaconda3/envs/rs3/lib/python3.8/site-packages/sklearn/base.py:310: UserWarning: Trying to unpickle estimator SimpleImputer from version 1.0.dev0 when using version 0.24.2. This might lead to breaking code or invalid results. Use at your own risk.
  warnings.warn(
/Users/pdeweird/opt/anaconda3/envs/rs3/lib/python3.8/site-packages/sklearn/base.py:310: UserWarning: Trying to unpickle estimator Pipeline from version 1.0.dev0 when using version 0.24.2. This might lead to breaking code or invalid results. Use at your own risk.
  warnings.warn(
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_text output_error">
<pre>
<span class="ansi-red-fg">---------------------------------------------------------------------------</span>
<span class="ansi-red-fg">AssertionError</span>                            Traceback (most recent call last)
<span class="ansi-green-fg">&lt;ipython-input-9-dc96c6fbd8d1&gt;</span> in <span class="ansi-cyan-fg">&lt;module&gt;</span>
<span class="ansi-green-intense-fg ansi-bold">      5</span>                                 conservatin_file<span class="ansi-blue-fg">=</span><span class="ansi-blue-fg">&#39;./data/target_data/conservation.pq&#39;</span><span class="ansi-blue-fg">,</span>
<span class="ansi-green-intense-fg ansi-bold">      6</span>                                 lite=False)
<span class="ansi-green-fg">----&gt; 7</span><span class="ansi-red-fg"> </span>pd<span class="ansi-blue-fg">.</span>testing<span class="ansi-blue-fg">.</span>assert_frame_equal<span class="ansi-blue-fg">(</span>scored_designs<span class="ansi-blue-fg">,</span> scored_designs_stored<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">      8</span> 

    <span class="ansi-red-fg">[... skipping hidden 2 frame]</span>

<span class="ansi-green-fg">pandas/_libs/testing.pyx</span> in <span class="ansi-cyan-fg">pandas._libs.testing.assert_almost_equal</span><span class="ansi-blue-fg">()</span>

<span class="ansi-green-fg">pandas/_libs/testing.pyx</span> in <span class="ansi-cyan-fg">pandas._libs.testing.assert_almost_equal</span><span class="ansi-blue-fg">()</span>

<span class="ansi-green-fg">~/opt/anaconda3/envs/rs3/lib/python3.8/site-packages/pandas/_testing.py</span> in <span class="ansi-cyan-fg">raise_assert_detail</span><span class="ansi-blue-fg">(obj, message, left, right, diff, index_values)</span>
<span class="ansi-green-intense-fg ansi-bold">   1071</span>         msg <span class="ansi-blue-fg">+=</span> <span class="ansi-blue-fg">f&#34;\n[diff]: {diff}&#34;</span>
<span class="ansi-green-intense-fg ansi-bold">   1072</span> 
<span class="ansi-green-fg">-&gt; 1073</span><span class="ansi-red-fg">     </span><span class="ansi-green-fg">raise</span> AssertionError<span class="ansi-blue-fg">(</span>msg<span class="ansi-blue-fg">)</span>
<span class="ansi-green-intense-fg ansi-bold">   1074</span> 
<span class="ansi-green-intense-fg ansi-bold">   1075</span> 

<span class="ansi-red-fg">AssertionError</span>: DataFrame.iloc[:, 64] (column name=&#34;Missing conservation information&#34;) are different

DataFrame.iloc[:, 64] (column name=&#34;Missing conservation information&#34;) values are different (1.5 %)
[index]: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, ...]
[left]:  [False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, ...]
[right]: [False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, ...]</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">activity_id_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;sgRNA Sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;sgRNA Context Sequence&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;Target Gene Symbol&#39;</span><span class="p">,</span> <span class="s1">&#39;Target Cut %&#39;</span><span class="p">]</span>
<span class="n">sanger_actvity_scores</span> <span class="o">=</span> <span class="p">(</span><span class="n">sanger_activity</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">scored_designs_stored</span><span class="p">,</span>
                                               <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span>
                                               <span class="n">on</span><span class="o">=</span><span class="n">activity_id_cols</span><span class="p">))</span>
<span class="n">gecko_activity_scores</span> <span class="o">=</span> <span class="p">(</span><span class="n">gecko_activity</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">scored_designs_stored</span><span class="p">,</span>
                                              <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">,</span>
                                              <span class="n">on</span><span class="o">=</span><span class="n">activity_id_cols</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="n">activity_col</span> <span class="o">=</span> <span class="s1">&#39;avg_mean_centered_neg_lfc&#39;</span>
<span class="n">score_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;RS3 Sequence Score (Hsu2013 tracr)&#39;</span><span class="p">,</span>
              <span class="s1">&#39;RS3 Sequence Score (Chen2013 tracr)&#39;</span><span class="p">,</span>
              <span class="s1">&#39;Target Score&#39;</span><span class="p">,</span>
              <span class="s1">&#39;RS3 Sequence (Hsu2013 tracr) + Target Score&#39;</span><span class="p">,</span>
              <span class="s1">&#39;RS3 Sequence (Chen2013 tracr) + Target Score&#39;</span><span class="p">]</span>
<span class="n">sanger_activity_cors</span> <span class="o">=</span> <span class="n">sanger_actvity_scores</span><span class="p">[[</span><span class="n">activity_col</span><span class="p">]</span> <span class="o">+</span> <span class="n">score_cols</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">()</span>
<span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">sanger_activity_cors</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="p">(</span><span class="n">sanger_activity_cors</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">activity_col</span><span class="p">,</span> <span class="s1">&#39;RS3 Sequence (Chen2013 tracr) + Target Score&#39;</span><span class="p">]</span> <span class="o">&gt;</span>
        <span class="n">sanger_activity_cors</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">activity_col</span><span class="p">,</span> <span class="s1">&#39;RS3 Sequence (Hsu2013 tracr) + Target Score&#39;</span><span class="p">]</span> <span class="o">&gt;</span>
        <span class="n">sanger_activity_cors</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">activity_col</span><span class="p">,</span> <span class="s1">&#39;RS3 Sequence Score (Chen2013 tracr)&#39;</span><span class="p">]</span> <span class="o">&gt;</span>
        <span class="n">sanger_activity_cors</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">activity_col</span><span class="p">,</span> <span class="s1">&#39;RS3 Sequence Score (Hsu2013 tracr)&#39;</span><span class="p">]</span> <span class="o">&gt;</span>
        <span class="n">sanger_activity_cors</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">activity_col</span><span class="p">,</span> <span class="s1">&#39;Target Score&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">gecko_activity_cors</span> <span class="o">=</span> <span class="n">gecko_activity_scores</span><span class="p">[[</span><span class="n">activity_col</span><span class="p">]</span> <span class="o">+</span> <span class="n">score_cols</span><span class="p">]</span><span class="o">.</span><span class="n">corr</span><span class="p">()</span>
<span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">gecko_activity_cors</span><span class="p">,</span> <span class="n">annot</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;RdBu_r&#39;</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mi">1</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="p">(</span><span class="n">gecko_activity_cors</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">activity_col</span><span class="p">,</span> <span class="s1">&#39;RS3 Sequence (Hsu2013 tracr) + Target Score&#39;</span><span class="p">]</span> <span class="o">&gt;</span>
        <span class="n">gecko_activity_cors</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">activity_col</span><span class="p">,</span> <span class="s1">&#39;RS3 Sequence Score (Hsu2013 tracr)&#39;</span><span class="p">]</span> <span class="o">&gt;</span>
        <span class="n">gecko_activity_cors</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">activity_col</span><span class="p">,</span> <span class="s1">&#39;RS3 Sequence Score (Chen2013 tracr)&#39;</span><span class="p">]</span> <span class="o">&gt;</span>
        <span class="n">gecko_activity_cors</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">activity_col</span><span class="p">,</span> <span class="s1">&#39;RS3 Sequence (Chen2013 tracr) + Target Score&#39;</span><span class="p">]</span> <span class="o">&gt;</span>
        <span class="n">gecko_activity_cors</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">activity_col</span><span class="p">,</span> <span class="s1">&#39;Target Score&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

</div>
 

