---

title: predicttarg


keywords: fastai
sidebar: home_sidebar

summary: "Rule set 3 target-site predictions"
description: "Rule set 3 target-site predictions"
nb_path: "03_predicttarg.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 03_predicttarg.ipynb
# command to build the docs after a change: nbdev_build_docs

-->
<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">lightgbm</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">rs3</span><span class="w"> </span><span class="kn">import</span> <span class="n">targetdata</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">scipy</span><span class="w"> </span><span class="kn">import</span> <span class="n">stats</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="vm">__file__</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="s1">''</span><span class="p">)</span> <span class="o">+</span> <span class="s1">'/03_predicttarg.ipynb'</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">multiprocessing</span>
<span class="n">max_n_jobs</span> <span class="o">=</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="load_target_model"><code>load_target_model</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/predicttarg.py#L11" style="float:right">[source]</a></h4>
<blockquote>
<p><code>load_target_model</code>(<strong><code>lite</code></strong>=<em><code>False</code></em>)</p>
</blockquote>
<p>Load rule set 3 target model</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="nb">type</span><span class="p">(</span><span class="n">load_target_model</span><span class="p">()[</span><span class="s1">'regressor'</span><span class="p">])</span> <span class="o">==</span> <span class="n">lightgbm</span><span class="o">.</span><span class="n">sklearn</span><span class="o">.</span><span class="n">LGBMRegressor</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/peterdeweirdt/miniforge3/envs/test_rs3_v6/lib/python3.9/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator SimpleImputer from version 1.0.dev0 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations
  warnings.warn(
/Users/peterdeweirdt/miniforge3/envs/test_rs3_v6/lib/python3.9/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator Pipeline from version 1.0.dev0 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations
  warnings.warn(
</pre>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="predict_target"><code>predict_target</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/predicttarg.py#L21" style="float:right">[source]</a></h4>
<blockquote>
<p><code>predict_target</code>(<strong><code>design_df</code></strong>, <strong><code>aa_subseq_df</code></strong>, <strong><code>domain_feature_df</code></strong>=<em><code>None</code></em>, <strong><code>conservation_feature_df</code></strong>=<em><code>None</code></em>, <strong><code>id_cols</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Make predictions using the Rule Set 3 target model. Note that if the protein_domain_df
or conservation_df are not supplied, then the lite model will be used, otherwise the full model is used.</p>
<p>:param design_df: DataFrame
:param aa_subseq_df: DataFrame
:param domain_feature_df: DataFrame
:param id_cols: list or str
:return: list</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">design_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s1">'test_data/sgrna-designs.txt'</span><span class="p">)</span>
<span class="n">design_targ_df</span> <span class="o">=</span> <span class="n">targetfeat</span><span class="o">.</span><span class="n">add_target_columns</span><span class="p">(</span><span class="n">design_df</span><span class="p">)</span>
<span class="n">id_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">'sgRNA Context Sequence'</span><span class="p">,</span> <span class="s1">'Target Cut Length'</span><span class="p">,</span> <span class="s1">'Target Transcript'</span><span class="p">,</span> <span class="s1">'Orientation'</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">aa_seq_df</span> <span class="o">=</span> <span class="n">targetdata</span><span class="o">.</span><span class="n">build_transcript_aa_seq_df</span><span class="p">(</span><span class="n">design_df</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">aa_subseq_df</span> <span class="o">=</span> <span class="n">targetfeat</span><span class="o">.</span><span class="n">get_aa_subseq_df</span><span class="p">(</span><span class="n">sg_designs</span><span class="o">=</span><span class="n">design_targ_df</span><span class="p">,</span> <span class="n">aa_seq_df</span><span class="o">=</span><span class="n">aa_seq_df</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
                                           <span class="n">id_cols</span><span class="o">=</span><span class="n">id_cols</span><span class="p">)</span>
<span class="n">aa_subseq_df</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting amino acid sequences
</pre>
</div>
</div>
<div class="output_area">
<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 4/4 [00:02&lt;00:00,  1.47it/s]
/Users/peterdeweirdt/Documents/rs3/rs3/targetdata.py:111: UserWarning: Unable to find translations for the following transcripts: ENST00000629399, ENST00000284049
  warnings.warn('Unable to find translations for the following transcripts: ' + ', '.join(missing_seqs))
</pre>
</div>
</div>
<div class="output_area">
<div class="output_html rendered_html output_subarea output_execute_result"><div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>Target Transcript</th>
<th>Target Total Length</th>
<th>Transcript Base</th>
<th>molecule</th>
<th>id</th>
<th>seq</th>
<th>version</th>
<th>desc</th>
<th>AA len</th>
<th>Target Cut Length</th>
<th>Orientation</th>
<th>sgRNA Context Sequence</th>
<th>AA Index</th>
<th>extended_seq</th>
<th>AA 0-Indexed</th>
<th>AA 0-Indexed padded</th>
<th>seq_start</th>
<th>seq_end</th>
<th>AA Subsequence</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>ENST00000259457.8</td>
<td>834</td>
<td>ENST00000259457</td>
<td>protein</td>
<td>ENSP00000259457</td>
<td>MAAVSVYAPPVGGFSFDNCRRNAVLEADFAKRGYKLPKVRKTGTTI...</td>
<td>3</td>
<td>None</td>
<td>277</td>
<td>191</td>
<td>sense</td>
<td>TGGAGCAGATACAAGAGCAACTGAAGGGAT</td>
<td>64</td>
<td>-----------------MAAVSVYAPPVGGFSFDNCRRNAVLEADF...</td>
<td>63</td>
<td>80</td>
<td>64</td>
<td>96</td>
<td>GVVYKDGIVLGADTRATEGMVVADKNCSKIHFI</td>
</tr>
<tr>
<th>1</th>
<td>ENST00000259457.8</td>
<td>834</td>
<td>ENST00000259457</td>
<td>protein</td>
<td>ENSP00000259457</td>
<td>MAAVSVYAPPVGGFSFDNCRRNAVLEADFAKRGYKLPKVRKTGTTI...</td>
<td>3</td>
<td>None</td>
<td>277</td>
<td>137</td>
<td>sense</td>
<td>CCGGAAAACTGGCACGACCATCGCTGGGGT</td>
<td>46</td>
<td>-----------------MAAVSVYAPPVGGFSFDNCRRNAVLEADF...</td>
<td>45</td>
<td>62</td>
<td>46</td>
<td>78</td>
<td>AKRGYKLPKVRKTGTTIAGVVYKDGIVLGADTR</td>
</tr>
<tr>
<th>2</th>
<td>ENST00000394249.8</td>
<td>1863</td>
<td>ENST00000394249</td>
<td>protein</td>
<td>ENSP00000377793</td>
<td>MRRSEVLAEESIVCLQKALNHLREIWELIGIPEDQRLQRTEVVKKH...</td>
<td>3</td>
<td>None</td>
<td>620</td>
<td>316</td>
<td>sense</td>
<td>TAGAAAAAGATTTGCGCACCCAAGTGGAAT</td>
<td>106</td>
<td>-----------------MRRSEVLAEESIVCLQKALNHLREIWELI...</td>
<td>105</td>
<td>122</td>
<td>106</td>
<td>138</td>
<td>EEGETTILQLEKDLRTQVELMRKQKKERKQELK</td>
</tr>
<tr>
<th>3</th>
<td>ENST00000394249.8</td>
<td>1863</td>
<td>ENST00000394249</td>
<td>protein</td>
<td>ENSP00000377793</td>
<td>MRRSEVLAEESIVCLQKALNHLREIWELIGIPEDQRLQRTEVVKKH...</td>
<td>3</td>
<td>None</td>
<td>620</td>
<td>787</td>
<td>antisense</td>
<td>TGGCCTTTGACCCAGACATAATGGTGGCCA</td>
<td>263</td>
<td>-----------------MRRSEVLAEESIVCLQKALNHLREIWELI...</td>
<td>262</td>
<td>279</td>
<td>263</td>
<td>295</td>
<td>WDRLQIPEEEREAVATIMSGSKAKVRKALQLEV</td>
</tr>
<tr>
<th>4</th>
<td>ENST00000361337.3</td>
<td>2298</td>
<td>ENST00000361337</td>
<td>protein</td>
<td>ENSP00000354522</td>
<td>MSGDHLHNDSQIEADFRLNDSHKHKDKHKDREHRHKEHKKEKDREK...</td>
<td>2</td>
<td>None</td>
<td>765</td>
<td>420</td>
<td>antisense</td>
<td>AAATACTCACTCATCCTCATCTCGAGGTCT</td>
<td>140</td>
<td>-----------------MSGDHLHNDSQIEADFRLNDSHKHKDKHK...</td>
<td>139</td>
<td>156</td>
<td>140</td>
<td>172</td>
<td>GYFVPPKEDIKPLKRPRDEDDADYKPKKIKTED</td>
</tr>
<tr>
<th>...</th>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr>
<th>391</th>
<td>ENST00000454402.7</td>
<td>1023</td>
<td>ENST00000454402</td>
<td>protein</td>
<td>ENSP00000408295</td>
<td>METSALKQQEQPAATKIRNLPWVEKYRPQTLNDLISHQDILSTIQK...</td>
<td>2</td>
<td>None</td>
<td>340</td>
<td>220</td>
<td>antisense</td>
<td>TGTCTTTATATAGCTGTTTCGCACAGGCTA</td>
<td>74</td>
<td>-----------------METSALKQQEQPAATKIRNLPWVEKYRPQ...</td>
<td>73</td>
<td>90</td>
<td>74</td>
<td>106</td>
<td>LYGPPGTGKTSTILACAKQLYKDKEFGSMVLEL</td>
</tr>
<tr>
<th>392</th>
<td>ENST00000254998.3</td>
<td>423</td>
<td>ENST00000254998</td>
<td>protein</td>
<td>ENSP00000254998</td>
<td>MASVDFKTYVDQACRAAEEFVNVYYTTMDKRRRLLSRLYMGTATLV...</td>
<td>2</td>
<td>None</td>
<td>140</td>
<td>79</td>
<td>sense</td>
<td>TTGTCAATGTCTACTACACCACCATGGATA</td>
<td>27</td>
<td>-----------------MASVDFKTYVDQACRAAEEFVNVYYTTMD...</td>
<td>26</td>
<td>43</td>
<td>27</td>
<td>59</td>
<td>DQACRAAEEFVNVYYTTMDKRRRLLSRLYMGTA</td>
</tr>
<tr>
<th>393</th>
<td>ENST00000254998.3</td>
<td>423</td>
<td>ENST00000254998</td>
<td>protein</td>
<td>ENSP00000254998</td>
<td>MASVDFKTYVDQACRAAEEFVNVYYTTMDKRRRLLSRLYMGTATLV...</td>
<td>2</td>
<td>None</td>
<td>140</td>
<td>115</td>
<td>sense</td>
<td>GGCGTTTGCTGTCCCGCCTGTACATGGGCA</td>
<td>39</td>
<td>-----------------MASVDFKTYVDQACRAAEEFVNVYYTTMD...</td>
<td>38</td>
<td>55</td>
<td>39</td>
<td>71</td>
<td>VYYTTMDKRRRLLSRLYMGTATLVWNGNAVSGQ</td>
</tr>
<tr>
<th>394</th>
<td>ENST00000381685.10</td>
<td>2067</td>
<td>ENST00000381685</td>
<td>protein</td>
<td>ENSP00000371101</td>
<td>MQVSSLNEVKIYSLSCGKSLPEWLSDRKKRALQKKDVDVRRRIELI...</td>
<td>5</td>
<td>None</td>
<td>688</td>
<td>776</td>
<td>antisense</td>
<td>ACTAGCAATGGCTTATCAGATCGAAGGTCA</td>
<td>259</td>
<td>-----------------MQVSSLNEVKIYSLSCGKSLPEWLSDRKK...</td>
<td>258</td>
<td>275</td>
<td>259</td>
<td>291</td>
<td>TMAVGTTTGQVLLYDLRSDKPLLVKDHQYGLPI</td>
</tr>
<tr>
<th>395</th>
<td>ENST00000381685.10</td>
<td>2067</td>
<td>ENST00000381685</td>
<td>protein</td>
<td>ENSP00000371101</td>
<td>MQVSSLNEVKIYSLSCGKSLPEWLSDRKKRALQKKDVDVRRRIELI...</td>
<td>5</td>
<td>None</td>
<td>688</td>
<td>322</td>
<td>sense</td>
<td>AAATTTTGTCTGATGACTACTCAAAGGTAT</td>
<td>108</td>
<td>-----------------MQVSSLNEVKIYSLSCGKSLPEWLSDRKK...</td>
<td>107</td>
<td>124</td>
<td>108</td>
<td>140</td>
<td>CLDSEVVTFEILSDDYSKIVFLHNDRYIEFHSQ</td>
</tr>
</tbody>
</table>
<p>396 rows × 19 columns</p>
</div></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">domain_df</span> <span class="o">=</span> <span class="n">targetdata</span><span class="o">.</span><span class="n">build_translation_overlap_df</span><span class="p">(</span><span class="n">aa_seq_df</span><span class="p">[</span><span class="s1">'id'</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">domain_feature_df</span> <span class="o">=</span> <span class="n">targetfeat</span><span class="o">.</span><span class="n">get_protein_domain_features</span><span class="p">(</span><span class="n">design_targ_df</span><span class="p">,</span> <span class="n">domain_df</span><span class="p">,</span> <span class="n">sources</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                                           <span class="n">id_cols</span><span class="o">=</span><span class="n">id_cols</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting protein domains
</pre>
</div>
</div>
<div class="output_area">
<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 198/198 [01:17&lt;00:00,  2.54it/s]
</pre>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">conservation_df</span> <span class="o">=</span> <span class="n">targetdata</span><span class="o">.</span><span class="n">build_conservation_df</span><span class="p">(</span><span class="n">design_df</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="n">max_n_jobs</span><span class="p">)</span>
<span class="n">conservation_feature_df</span> <span class="o">=</span> <span class="n">targetfeat</span><span class="o">.</span><span class="n">get_conservation_features</span><span class="p">(</span><span class="n">design_targ_df</span><span class="p">,</span> <span class="n">conservation_df</span><span class="p">,</span>
                                                             <span class="n">small_width</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">large_width</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
                                                             <span class="n">conservation_column</span><span class="o">=</span><span class="s1">'ranked_conservation'</span><span class="p">,</span>
                                                             <span class="n">id_cols</span><span class="o">=</span><span class="n">id_cols</span><span class="p">)</span>
<span class="n">conservation_feature_df</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting conservation
</pre>
</div>
</div>
<div class="output_area">
<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 200/200 [04:30&lt;00:00,  1.35s/it]
/Users/peterdeweirdt/Documents/rs3/rs3/targetdata.py:346: UserWarning: Failed to get conservation scores for 2 transcriptsENST00000629399, ENST00000284049
  warnings.warn('Failed to get conservation scores for ' + str(len(failed_list)) +
</pre>
</div>
</div>
<div class="output_area">
<div class="output_html rendered_html output_subarea output_execute_result"><div>
<style scoped="">
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
<thead>
<tr style="text-align: right;">
<th></th>
<th>sgRNA Context Sequence</th>
<th>Target Cut Length</th>
<th>Target Transcript</th>
<th>Orientation</th>
<th>cons_4</th>
<th>cons_32</th>
</tr>
</thead>
<tbody>
<tr>
<th>0</th>
<td>AAAAGAATGATGAAAAGACACCACAGGGAG</td>
<td>244</td>
<td>ENST00000610426.5</td>
<td>sense</td>
<td>0.218231</td>
<td>0.408844</td>
</tr>
<tr>
<th>1</th>
<td>AAAAGAGCCATGAATCTAAACATCAGGAAT</td>
<td>640</td>
<td>ENST00000223073.6</td>
<td>sense</td>
<td>0.129825</td>
<td>0.278180</td>
</tr>
<tr>
<th>2</th>
<td>AAAAGCGCCAAATGGCCCGAGAATTGGGAG</td>
<td>709</td>
<td>ENST00000331923.9</td>
<td>sense</td>
<td>0.470906</td>
<td>0.532305</td>
</tr>
<tr>
<th>3</th>
<td>AAACAGAAAAAGTTAAAATCACCAAGGTGT</td>
<td>496</td>
<td>ENST00000283882.4</td>
<td>sense</td>
<td>0.580556</td>
<td>0.602708</td>
</tr>
<tr>
<th>4</th>
<td>AAACAGATGGAAGATGCTTACCGGGGGACC</td>
<td>132</td>
<td>ENST00000393047.8</td>
<td>sense</td>
<td>0.283447</td>
<td>0.414293</td>
</tr>
<tr>
<th>...</th>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr>
<th>391</th>
<td>TTTGATTGCATTAAGGTTGGACTCTGGATT</td>
<td>246</td>
<td>ENST00000249269.9</td>
<td>sense</td>
<td>0.580612</td>
<td>0.618707</td>
</tr>
<tr>
<th>392</th>
<td>TTTGCCCACAGCTCCAAAGCATCGCGGAGA</td>
<td>130</td>
<td>ENST00000227618.8</td>
<td>sense</td>
<td>0.323770</td>
<td>0.416368</td>
</tr>
<tr>
<th>393</th>
<td>TTTTACAGTGCGATGTATGATGTATGGCTT</td>
<td>119</td>
<td>ENST00000338366.6</td>
<td>sense</td>
<td>0.788000</td>
<td>0.537417</td>
</tr>
<tr>
<th>394</th>
<td>TTTTGGATCTCGTAGTGATTCAAGAGGGAA</td>
<td>233</td>
<td>ENST00000629496.3</td>
<td>sense</td>
<td>0.239630</td>
<td>0.347615</td>
</tr>
<tr>
<th>395</th>
<td>TTTTTGTTACTACAGGTTCGCTGCTGGGAA</td>
<td>201</td>
<td>ENST00000395840.6</td>
<td>sense</td>
<td>0.693767</td>
<td>0.639044</td>
</tr>
</tbody>
</table>
<p>396 rows × 6 columns</p>
</div></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">predictions</span> <span class="o">=</span> <span class="n">predict_target</span><span class="p">(</span><span class="n">design_df</span><span class="o">=</span><span class="n">design_df</span><span class="p">,</span>
                             <span class="n">aa_subseq_df</span><span class="o">=</span><span class="n">aa_subseq_df</span><span class="p">,</span>
                             <span class="n">domain_feature_df</span><span class="o">=</span><span class="n">domain_feature_df</span><span class="p">,</span>
                             <span class="n">conservation_feature_df</span><span class="o">=</span><span class="n">conservation_feature_df</span><span class="p">)</span>
<span class="n">design_df</span><span class="p">[</span><span class="s1">'Target Score'</span><span class="p">]</span> <span class="o">=</span> <span class="n">predictions</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/peterdeweirdt/miniforge3/envs/test_rs3_v6/lib/python3.9/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator SimpleImputer from version 1.0.dev0 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations
  warnings.warn(
/Users/peterdeweirdt/miniforge3/envs/test_rs3_v6/lib/python3.9/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator Pipeline from version 1.0.dev0 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations
  warnings.warn(
/Users/peterdeweirdt/miniforge3/envs/test_rs3_v6/lib/python3.9/site-packages/sklearn/base.py:443: UserWarning: X has feature names, but SimpleImputer was fitted without feature names
  warnings.warn(
</pre>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">lite_predictions</span> <span class="o">=</span> <span class="n">predict_target</span><span class="p">(</span><span class="n">design_df</span><span class="o">=</span><span class="n">design_df</span><span class="p">,</span>
                                  <span class="n">aa_subseq_df</span><span class="o">=</span><span class="n">aa_subseq_df</span><span class="p">)</span>
<span class="n">design_df</span><span class="p">[</span><span class="s1">'Target Score Lite'</span><span class="p">]</span> <span class="o">=</span> <span class="n">lite_predictions</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stderr output_text">
<pre>/Users/peterdeweirdt/miniforge3/envs/test_rs3_v6/lib/python3.9/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator SimpleImputer from version 1.0.dev0 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations
  warnings.warn(
/Users/peterdeweirdt/miniforge3/envs/test_rs3_v6/lib/python3.9/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator Pipeline from version 1.0.dev0 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations
  warnings.warn(
/Users/peterdeweirdt/miniforge3/envs/test_rs3_v6/lib/python3.9/site-packages/sklearn/base.py:443: UserWarning: X has feature names, but SimpleImputer was fitted without feature names
  warnings.warn(
</pre>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="n">stats</span><span class="o">.</span><span class="n">pearsonr</span><span class="p">(</span><span class="n">design_df</span><span class="p">[</span><span class="s1">'Target Score'</span><span class="p">],</span> <span class="n">design_df</span><span class="p">[</span><span class="s1">'Target Score Lite'</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.5</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">sanger_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">'test_data/Behan2019_activity.csv'</span><span class="p">)</span>
<span class="n">gecko_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">'test_data/Aguirre2016_activity.csv'</span><span class="p">)</span>

<span class="n">sanger_designs</span> <span class="o">=</span> <span class="n">sanger_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">design_df</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">'inner'</span><span class="p">,</span>
                                 <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">'sgRNA Sequence'</span><span class="p">,</span> <span class="s1">'sgRNA Context Sequence'</span><span class="p">,</span> <span class="s1">'Target Gene Symbol'</span><span class="p">,</span>
                                     <span class="s1">'Target Cut %'</span><span class="p">])</span>
<span class="n">gecko_designs</span> <span class="o">=</span> <span class="n">gecko_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">design_df</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">'inner'</span><span class="p">,</span>
                                <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">'sgRNA Sequence'</span><span class="p">,</span> <span class="s1">'sgRNA Context Sequence'</span><span class="p">,</span> <span class="s1">'Target Gene Symbol'</span><span class="p">,</span>
                                    <span class="s1">'Target Cut %'</span><span class="p">])</span>
<span class="k">assert</span> <span class="n">stats</span><span class="o">.</span><span class="n">pearsonr</span><span class="p">(</span><span class="n">sanger_designs</span><span class="p">[</span><span class="s1">'avg_mean_centered_neg_lfc'</span><span class="p">],</span>
                      <span class="n">sanger_designs</span><span class="p">[</span><span class="s1">'Target Score'</span><span class="p">])[</span><span class="mi">0</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.1</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

</div>
