---

title: targetfeat


keywords: fastai
sidebar: home_sidebar

summary: "Module to generate target site features"
description: "Module to generate target site features"
nb_path: "02_targetfeat.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 02_targetfeat.ipynb
# command to build the docs after a change: nbdev_build_docs

-->
<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">rs3</span><span class="w"> </span><span class="kn">import</span> <span class="n">targetdata</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">multiprocessing</span>
<span class="n">max_n_jobs</span> <span class="o">=</span> <span class="n">multiprocessing</span><span class="o">.</span><span class="n">cpu_count</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="add_target_columns"><code>add_target_columns</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L14" style="float:right">[source]</a></h4>
<blockquote>
<p><code>add_target_columns</code>(<strong><code>design_df</code></strong>, <strong><code>transcript_id_col</code></strong>=<em><code>'Target Transcript'</code></em>, <strong><code>cut_pos_col</code></strong>=<em><code>'Target Cut Length'</code></em>, <strong><code>transcript_base_col</code></strong>=<em><code>'Transcript Base'</code></em>)</p>
</blockquote>
<p>Add ['AA Index' and 'Transcript Base'] to design df</p>
<p>:param design_df: DataFrame
:return: DataFrame</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">design_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s1">'test_data/sgrna-designs.txt'</span><span class="p">)</span>
<span class="n">design_targ_df</span> <span class="o">=</span> <span class="n">add_target_columns</span><span class="p">(</span><span class="n">design_df</span><span class="p">)</span>
<span class="k">assert</span> <span class="s1">'AA Index'</span> <span class="ow">in</span> <span class="n">design_targ_df</span><span class="o">.</span><span class="n">columns</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Position-Features">Position Features<a class="anchor-link" href="#Position-Features"> </a></h2><p>The first feature class we consider is where the guide targets within the annotated transcript</p>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="get_position_features"><code>get_position_features</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L28" style="float:right">[source]</a></h4>
<blockquote>
<p><code>get_position_features</code>(<strong><code>sg_df</code></strong>, <strong><code>id_cols</code></strong>)</p>
</blockquote>
<p>Get  features ['Target Cut %', 'sense']</p>
<p>:param sg_df: DataFrame
:param id_cols: list
:return: DataFrame</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Amino-Acid-Features">Amino Acid Features<a class="anchor-link" href="#Amino-Acid-Features"> </a></h2><p>We calculate a set of features from the amino acid sequence around the cutsite itself</p>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">aas</span> <span class="o">=</span> <span class="p">[</span><span class="s1">'A'</span><span class="p">,</span> <span class="s1">'C'</span><span class="p">,</span> <span class="s1">'D'</span><span class="p">,</span> <span class="s1">'E'</span><span class="p">,</span> <span class="s1">'F'</span><span class="p">,</span>
       <span class="s1">'G'</span><span class="p">,</span> <span class="s1">'H'</span><span class="p">,</span> <span class="s1">'I'</span><span class="p">,</span> <span class="s1">'K'</span><span class="p">,</span> <span class="s1">'L'</span><span class="p">,</span>
       <span class="s1">'M'</span><span class="p">,</span> <span class="s1">'N'</span><span class="p">,</span> <span class="s1">'P'</span><span class="p">,</span> <span class="s1">'Q'</span><span class="p">,</span> <span class="s1">'R'</span><span class="p">,</span>
       <span class="s1">'S'</span><span class="p">,</span> <span class="s1">'T'</span><span class="p">,</span> <span class="s1">'V'</span><span class="p">,</span> <span class="s1">'W'</span><span class="p">,</span> <span class="s1">'Y'</span><span class="p">,</span> <span class="s1">'*'</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="get_one_aa_frac"><code>get_one_aa_frac</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L40" style="float:right">[source]</a></h4>
<blockquote>
<p><code>get_one_aa_frac</code>(<strong><code>feature_dict</code></strong>, <strong><code>aa_sequence</code></strong>, <strong><code>aas</code></strong>)</p>
</blockquote>
<p>Get fraction of single aa</p>
<p>:param feature_dict: dict, feature dictionary
:param aa_sequence: str, amino acid sequence
:param aas: list, list of amino acids</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">one_aa_ft</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">get_one_aa_frac</span><span class="p">(</span><span class="n">one_aa_ft</span><span class="p">,</span> <span class="s1">'ACDG*-'</span><span class="p">,</span> <span class="n">aas</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">one_aa_ft</span><span class="p">[</span><span class="s1">'A'</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="o">/</span><span class="mi">6</span>
<span class="k">assert</span> <span class="n">one_aa_ft</span><span class="p">[</span><span class="s1">'Q'</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="get_aa_aromaticity"><code>get_aa_aromaticity</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L52" style="float:right">[source]</a></h4>
<blockquote>
<p><code>get_aa_aromaticity</code>(<strong><code>feature_dict</code></strong>, <strong><code>analyzed_seq</code></strong>)</p>
</blockquote>
<p>Get fraction of aromatic amino acids in a sequence.</p>
<p>Phe (F) + Trp (W) + Tyr (Y)</p>
<p>:param feature_dict:
:param analyzed_seq: ProteinAnalysis object</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="get_aa_hydrophobicity"><code>get_aa_hydrophobicity</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L63" style="float:right">[source]</a></h4>
<blockquote>
<p><code>get_aa_hydrophobicity</code>(<strong><code>feature_dict</code></strong>, <strong><code>analyzed_seq</code></strong>)</p>
</blockquote>
<p>Grand Average of Hydropathy</p>
<p>The GRAVY value is calculated by adding the hydropathy value for each residue and dividing
by the length of the sequence (Kyte and Doolittle; 1982). The larger the number, the more hydrophobic the
amino acid</p>
<p>:param feature_dict: dict
:param analyzed_seq: ProteinAnalysis object</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="get_aa_ip"><code>get_aa_ip</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L76" style="float:right">[source]</a></h4>
<blockquote>
<p><code>get_aa_ip</code>(<strong><code>feature_dict</code></strong>, <strong><code>analyzed_seq</code></strong>)</p>
</blockquote>
<p>Get the Isoelectric Point of an amino acid sequence</p>
<p>Charge of amino acid</p>
<p>:param feature_dict: dict
:param analyzed_seq: ProteinAnalysis object</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="get_aa_secondary_structure"><code>get_aa_secondary_structure</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L87" style="float:right">[source]</a></h4>
<blockquote>
<p><code>get_aa_secondary_structure</code>(<strong><code>feature_dict</code></strong>, <strong><code>analyzed_seq</code></strong>)</p>
</blockquote>
<p>Get the fraction of amion acids that tend to be in a helix, turn or sheet</p>
<p>:param feature_dict: dict
:param analyzed_seq: ProteinAnalysis object</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">aa_biochemical_fts1</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">get_aa_aromaticity</span><span class="p">(</span><span class="n">aa_biochemical_fts1</span><span class="p">,</span> <span class="n">ProteinAnalysis</span><span class="p">(</span><span class="s1">'FWYA'</span><span class="p">))</span>
<span class="n">aa_biochemical_fts2</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">get_aa_aromaticity</span><span class="p">(</span><span class="n">aa_biochemical_fts2</span><span class="p">,</span> <span class="n">ProteinAnalysis</span><span class="p">(</span><span class="s1">'AAAA'</span><span class="p">))</span>
<span class="k">assert</span> <span class="n">aa_biochemical_fts1</span><span class="p">[</span><span class="s1">'Aromaticity'</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">aa_biochemical_fts2</span><span class="p">[</span><span class="s1">'Aromaticity'</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="featurize_aa_seqs"><code>featurize_aa_seqs</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L97" style="float:right">[source]</a></h4>
<blockquote>
<p><code>featurize_aa_seqs</code>(<strong><code>aa_sequences</code></strong>, <strong><code>features</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Get feature DataFrame for a list of amino acid sequences</p>
<p>:param aa_sequences: list of str
:param features: list or None
:return: DataFrame</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">ft_dict_df</span> <span class="o">=</span> <span class="n">featurize_aa_seqs</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="s1">'ACDG*-'</span><span class="p">,</span> <span class="s1">'CDG*--'</span><span class="p">,</span> <span class="s1">'LLLLLL'</span><span class="p">]))</span>
<span class="k">assert</span> <span class="n">ft_dict_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">'LLLLLL'</span><span class="p">,</span> <span class="s1">'Hydrophobicity'</span><span class="p">]</span> <span class="o">==</span> <span class="n">ft_dict_df</span><span class="p">[</span><span class="s1">'Hydrophobicity'</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="extract_amino_acid_subsequence"><code>extract_amino_acid_subsequence</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L132" style="float:right">[source]</a></h4>
<blockquote>
<p><code>extract_amino_acid_subsequence</code>(<strong><code>sg_aas</code></strong>, <strong><code>width</code></strong>)</p>
</blockquote>
<p>Get the amino acid subsequence with a width of <code>width</code> on either side of the Amino Acid index</p>
<p>:param sg_aas: DataFrame, sgRNA designs merged with amino acid sequence
:param width: int
:return: DataFrame</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">small_aa_seq_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s1">'AA Index'</span><span class="p">:</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">9</span><span class="p">],</span>
                                    <span class="s1">'seq'</span><span class="p">:</span> <span class="p">[</span><span class="s1">'MAVLKYSLW'</span><span class="p">]</span><span class="o">*</span><span class="mi">3</span><span class="p">})</span>
<span class="n">small_aa_subseq_df</span> <span class="o">=</span> <span class="n">extract_amino_acid_subsequence</span><span class="p">(</span><span class="n">small_aa_seq_df</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="n">actual_subseqs</span> <span class="o">=</span> <span class="n">small_aa_subseq_df</span><span class="p">[</span><span class="s1">'AA Subsequence'</span><span class="p">]</span>
<span class="n">expected_subseqs</span> <span class="o">=</span> <span class="p">[</span><span class="s1">'--MAV'</span><span class="p">,</span> <span class="s1">'VLKYS'</span><span class="p">,</span> <span class="s1">'SLW*-'</span><span class="p">]</span>
<span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">actual_subseqs</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">expected_subseqs</span><span class="p">)</span>
<span class="k">assert</span> <span class="nb">all</span><span class="p">([</span><span class="n">a</span> <span class="o">==</span> <span class="n">b</span> <span class="k">for</span> <span class="n">a</span><span class="p">,</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">actual_subseqs</span><span class="p">,</span> <span class="n">expected_subseqs</span><span class="p">)])</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="get_aa_subseq_df"><code>get_aa_subseq_df</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L155" style="float:right">[source]</a></h4>
<blockquote>
<p><code>get_aa_subseq_df</code>(<strong><code>sg_designs</code></strong>, <strong><code>aa_seq_df</code></strong>, <strong><code>width</code></strong>, <strong><code>id_cols</code></strong>, <strong><code>transcript_base_col</code></strong>=<em><code>'Transcript Base'</code></em>, <strong><code>target_transcript_col</code></strong>=<em><code>'Target Transcript'</code></em>, <strong><code>aa_index_col</code></strong>=<em><code>'AA Index'</code></em>)</p>
</blockquote>
<p>Get the amino acid subsequences for a design dataframe</p>
<p>:param sg_designs: DataFrame
:param aa_seq_df: DataFrame, Transcript Base and (AA) seq
:param width: int, length on each side of the cut site
:param transcript_base_col: str
:param target_transcript_col: str
:param aa_index_col: str
:return: DataFrame</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">aa_seq_df</span> <span class="o">=</span> <span class="n">targetdata</span><span class="o">.</span><span class="n">build_transcript_aa_seq_df</span><span class="p">(</span><span class="n">design_targ_df</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting amino acid sequences
</pre>
</div>
</div>
<div class="output_area">
<div class="output_subarea output_stream output_stderr output_text">
<pre>  0%|          | 0/4 [00:00&lt;?, ?it/s]</pre>
</div>
</div>
<div class="output_area">
<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 4/4 [00:02&lt;00:00,  1.56it/s]
/Users/peterdeweirdt/Documents/rs3/rs3/targetdata.py:110: UserWarning: Unable to find translations for the following transcripts: ENST00000629399, ENST00000284049
  warnings.warn('Unable to find translations for the following transcripts: ' + ', '.join(missing_seqs))
</pre>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">aa_subseq_df</span> <span class="o">=</span> <span class="n">get_aa_subseq_df</span><span class="p">(</span><span class="n">sg_designs</span><span class="o">=</span><span class="n">design_targ_df</span><span class="p">,</span> <span class="n">aa_seq_df</span><span class="o">=</span><span class="n">aa_seq_df</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
                                <span class="n">id_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">'sgRNA Context Sequence'</span><span class="p">,</span> <span class="s1">'Target Cut Length'</span><span class="p">,</span> <span class="s1">'Target Transcript'</span><span class="p">,</span> <span class="s1">'Orientation'</span><span class="p">])</span>
<span class="k">assert</span> <span class="p">(</span><span class="n">aa_subseq_df</span><span class="p">[</span><span class="s1">'AA Subsequence'</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">len</span><span class="p">()</span> <span class="o">==</span> <span class="mi">33</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
<span class="k">assert</span> <span class="nb">abs</span><span class="p">(</span><span class="n">aa_subseq_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="n">design_targ_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mi">10</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">codon_map_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">'test_data/codon_map.csv'</span><span class="p">)</span>

<span class="k">def</span><span class="w"> </span><span class="nf">get_rev_comp</span><span class="p">(</span><span class="n">sgrna</span><span class="p">):</span>
<span class="w">    </span><span class="sd">"""Get reverse compliment of a guide"""</span>
    <span class="n">nt_map</span> <span class="o">=</span> <span class="p">{</span><span class="s1">'A'</span><span class="p">:</span> <span class="s1">'T'</span><span class="p">,</span> <span class="s1">'T'</span><span class="p">:</span> <span class="s1">'A'</span><span class="p">,</span> <span class="s1">'G'</span><span class="p">:</span> <span class="s1">'C'</span><span class="p">,</span> <span class="s1">'C'</span><span class="p">:</span> <span class="s1">'G'</span><span class="p">}</span>
    <span class="n">rev_comp</span> <span class="o">=</span> <span class="s1">''</span>
    <span class="k">for</span> <span class="n">nt</span> <span class="ow">in</span> <span class="n">sgrna</span><span class="p">:</span>
        <span class="n">rev_comp</span> <span class="o">+=</span> <span class="n">nt_map</span><span class="p">[</span><span class="n">nt</span><span class="p">]</span>
    <span class="n">rev_comp</span> <span class="o">=</span> <span class="n">rev_comp</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">rev_comp</span>

<span class="n">codon_map</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">codon_map_df</span><span class="p">[</span><span class="s1">'Amino Acid'</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">codon_map_df</span><span class="p">[</span><span class="s1">'Codon'</span><span class="p">])</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
<span class="n">row</span> <span class="o">=</span> <span class="n">aa_subseq_df</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:]</span>
<span class="n">subseq</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">'AA Subsequence'</span><span class="p">]</span>
<span class="n">context</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">'sgRNA Context Sequence'</span><span class="p">]</span>
<span class="n">rc_context</span> <span class="o">=</span> <span class="n">get_rev_comp</span><span class="p">(</span><span class="n">context</span><span class="p">)</span>
<span class="n">translations</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="n">rc_translations</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]:</span>
    <span class="n">translations</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="s1">''</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">codon_map</span><span class="p">[</span><span class="n">context</span><span class="p">[</span><span class="n">j</span><span class="p">:</span><span class="n">j</span><span class="o">+</span><span class="mi">3</span><span class="p">]]</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">context</span><span class="p">),</span> <span class="mi">3</span><span class="p">)</span>
                               <span class="k">if</span> <span class="p">(</span><span class="n">j</span> <span class="o">+</span> <span class="mi">3</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">context</span><span class="p">)])</span>
    <span class="n">rc_translations</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="s1">''</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">codon_map</span><span class="p">[</span><span class="n">rc_context</span><span class="p">[</span><span class="n">j</span><span class="p">:</span><span class="n">j</span><span class="o">+</span><span class="mi">3</span><span class="p">]]</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">rc_context</span><span class="p">),</span> <span class="mi">3</span><span class="p">)</span>
                                  <span class="k">if</span> <span class="p">(</span><span class="n">j</span> <span class="o">+</span> <span class="mi">3</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rc_context</span><span class="p">)])</span>
<span class="k">assert</span> <span class="p">((</span><span class="n">translations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">translations</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">translations</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">)</span> <span class="ow">or</span>
        <span class="p">(</span><span class="n">rc_translations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">rc_translations</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">rc_translations</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="get_amino_acid_features"><code>get_amino_acid_features</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L177" style="float:right">[source]</a></h4>
<blockquote>
<p><code>get_amino_acid_features</code>(<strong><code>aa_subseq_df</code></strong>, <strong><code>features</code></strong>, <strong><code>id_cols</code></strong>)</p>
</blockquote>
<p>Featurize amino acid sequences</p>
<p>:param aa_subseq_df: DataFrame
:param features: list
:param id_cols: list
:return: DataFrame</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">aa_features</span> <span class="o">=</span> <span class="n">get_amino_acid_features</span><span class="p">(</span><span class="n">aa_subseq_df</span><span class="o">=</span><span class="n">aa_subseq_df</span><span class="p">,</span>
                                      <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s1">'Pos. Ind. 1mer'</span><span class="p">,</span>
                                                <span class="s1">'Hydrophobicity'</span><span class="p">,</span> <span class="s1">'Aromaticity'</span><span class="p">,</span>
                                                <span class="s1">'Isoelectric Point'</span><span class="p">,</span> <span class="s1">'Secondary Structure'</span><span class="p">],</span>
                                      <span class="n">id_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">'sgRNA Context Sequence'</span><span class="p">,</span> <span class="s1">'Target Cut Length'</span><span class="p">,</span>
                                               <span class="s1">'Target Transcript'</span><span class="p">,</span> <span class="s1">'Orientation'</span><span class="p">])</span>
<span class="k">assert</span> <span class="n">aa_features</span><span class="p">[</span><span class="s1">'L'</span><span class="p">]</span><span class="o">.</span><span class="n">idxmax</span><span class="p">()</span> <span class="o">==</span> <span class="n">aa_features</span><span class="p">[</span><span class="s1">'Hydrophobicity'</span><span class="p">]</span><span class="o">.</span><span class="n">idxmax</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Protein-Domain-Features">Protein Domain Features<a class="anchor-link" href="#Protein-Domain-Features"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="get_protein_domain_features"><code>get_protein_domain_features</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L202" style="float:right">[source]</a></h4>
<blockquote>
<p><code>get_protein_domain_features</code>(<strong><code>sg_design_df</code></strong>, <strong><code>protein_domains</code></strong>, <strong><code>id_cols</code></strong>, <strong><code>sources</code></strong>=<em><code>None</code></em>, <strong><code>transcript_base_col</code></strong>=<em><code>'Transcript Base'</code></em>, <strong><code>aa_index_col</code></strong>=<em><code>'AA Index'</code></em>, <strong><code>domain_type_col</code></strong>=<em><code>'type'</code></em>, <strong><code>domain_start_col</code></strong>=<em><code>'start'</code></em>, <strong><code>domain_end_col</code></strong>=<em><code>'end'</code></em>)</p>
</blockquote>
<p>Get binary dataframe of protein domains</p>
<p>:param sg_design_df: DataFrame, with columns [transcript_base_col, aa_index_col]
:param protein_domains: DataFrame, with columns [transcript_base_col, domain_type_col]
:param id_cols: list
:param sources: list. list of database types to include
:param transcript_base_col: str
:param aa_index_col: str
:param domain_type_col: str
:param domain_start_col: str
:param domain_end_col: str
:return: DataFrame, with binary features for protein domains</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">domain_df</span> <span class="o">=</span> <span class="n">targetdata</span><span class="o">.</span><span class="n">build_translation_overlap_df</span><span class="p">(</span><span class="n">aa_seq_df</span><span class="p">[</span><span class="s1">'id'</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">protein_domain_feature_df</span> <span class="o">=</span> <span class="n">get_protein_domain_features</span><span class="p">(</span><span class="n">design_targ_df</span><span class="p">,</span> <span class="n">domain_df</span><span class="p">,</span> <span class="n">sources</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                                        <span class="n">id_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">'sgRNA Context Sequence'</span><span class="p">,</span> <span class="s1">'Target Cut Length'</span><span class="p">,</span>
                                                                 <span class="s1">'AA Index'</span><span class="p">,</span> <span class="s1">'Target Transcript'</span><span class="p">,</span> <span class="s1">'Orientation'</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting protein domains
</pre>
</div>
</div>
<div class="output_area">
<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 198/198 [00:50&lt;00:00,  3.92it/s]
</pre>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="n">protein_domain_feature_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein_domain_feature_df</span><span class="p">[</span><span class="s1">'sgRNA Context Sequence'</span><span class="p">]</span> <span class="o">==</span> <span class="s1">'AAAAGAGCCATGAATCTAAACATCAGGAAT'</span><span class="p">,</span>
                                     <span class="p">[</span><span class="s1">'PANTHER'</span><span class="p">,</span> <span class="s1">'ncoils'</span><span class="p">,</span> <span class="s1">'Seg'</span><span class="p">,</span> <span class="s1">'MobiDBLite'</span><span class="p">]]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">4</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Conservation-Features">Conservation Features<a class="anchor-link" href="#Conservation-Features"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="get_conservation_ranges"><code>get_conservation_ranges</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L247" style="float:right">[source]</a></h4>
<blockquote>
<p><code>get_conservation_ranges</code>(<strong><code>cut_pos</code></strong>, <strong><code>small_width</code></strong>, <strong><code>large_width</code></strong>)</p>
</blockquote>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="get_conservation_features"><code>get_conservation_features</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L253" style="float:right">[source]</a></h4>
<blockquote>
<p><code>get_conservation_features</code>(<strong><code>sg_designs</code></strong>, <strong><code>conservation_df</code></strong>, <strong><code>conservation_column</code></strong>, <strong><code>small_width</code></strong>, <strong><code>large_width</code></strong>, <strong><code>id_cols</code></strong>)</p>
</blockquote>
<p>Get conservation features</p>
<p>:param sg_designs: DataFrame
:param conservation_df: DataFrame, tidy conservation scores indexed by Transcript Base and target position
:param conservation_column: str, name of column to calculate scores with
:param small_width: int, small window length to average scores in one direction
:param large_width: int, large window length to average scores in the one direction
:return: DataFrame of conservation features</p>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">conservation_df</span> <span class="o">=</span> <span class="n">targetdata</span><span class="o">.</span><span class="n">build_conservation_df</span><span class="p">(</span><span class="n">design_targ_df</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="n">max_n_jobs</span><span class="p">)</span>
<span class="n">conservation_features</span> <span class="o">=</span> <span class="n">get_conservation_features</span><span class="p">(</span><span class="n">design_targ_df</span><span class="p">,</span> <span class="n">conservation_df</span><span class="p">,</span>
                                                  <span class="n">small_width</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">large_width</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
                                                  <span class="n">conservation_column</span><span class="o">=</span><span class="s1">'ranked_conservation'</span><span class="p">,</span>
                                                  <span class="n">id_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">'sgRNA Context Sequence'</span><span class="p">,</span> <span class="s1">'Target Cut Length'</span><span class="p">,</span>
                                                           <span class="s1">'Target Transcript'</span><span class="p">,</span> <span class="s1">'Orientation'</span><span class="p">])</span>
<span class="n">merged_features</span> <span class="o">=</span> <span class="n">protein_domain_feature_df</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">conservation_features</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">'inner'</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">'sgRNA Context Sequence'</span><span class="p">,</span>
                                                                                          <span class="s1">'Target Cut Length'</span><span class="p">,</span>
                                                                                          <span class="s1">'Target Transcript'</span><span class="p">,</span>
                                                                                          <span class="s1">'Orientation'</span><span class="p">])</span>
<span class="n">smart_avg_cons</span> <span class="o">=</span> <span class="n">merged_features</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">merged_features</span><span class="p">[</span><span class="s1">'Smart'</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">),</span> <span class="s1">'cons_32'</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="n">non_smart_avg_cons</span> <span class="o">=</span> <span class="n">merged_features</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">merged_features</span><span class="p">[</span><span class="s1">'Smart'</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">),</span> <span class="s1">'cons_32'</span><span class="p">]</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="k">assert</span> <span class="n">smart_avg_cons</span> <span class="o">&gt;</span> <span class="n">non_smart_avg_cons</span>
</pre></div>
</div>
</div>
</div>
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting conservation
</pre>
</div>
</div>
<div class="output_area">
<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 200/200 [04:56&lt;00:00,  1.48s/it]
/Users/peterdeweirdt/Documents/rs3/rs3/targetdata.py:345: UserWarning: Failed to get conservation scores for 2 transcriptsENST00000629399, ENST00000284049
  warnings.warn('Failed to get conservation scores for ' + str(len(failed_list)) +
</pre>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Combining-target-features">Combining target features<a class="anchor-link" href="#Combining-target-features"> </a></h2><p>We'll combine, the position, amino acid and domain feature matrices into a single target feature matrix</p>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="output_wrapper">
<div class="output">
<div class="output_area">
<div class="output_markdown rendered_html output_subarea">
<h4 class="doc_header" id="merge_feature_dfs"><code>merge_feature_dfs</code><a class="source_link" href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L292" style="float:right">[source]</a></h4>
<blockquote>
<p><code>merge_feature_dfs</code>(<strong><code>design_df</code></strong>, <strong><code>aa_subseq_df</code></strong>, <strong><code>aa_features</code></strong>=<em><code>None</code></em>, <strong><code>domain_df</code></strong>=<em><code>None</code></em>, <strong><code>conservation_df</code></strong>=<em><code>None</code></em>, <strong><code>id_cols</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
</div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="n">feature_df</span><span class="p">,</span> <span class="n">feature_list</span> <span class="o">=</span> <span class="n">merge_feature_dfs</span><span class="p">(</span><span class="n">design_df</span><span class="o">=</span><span class="n">design_df</span><span class="p">,</span>
                                             <span class="n">aa_subseq_df</span><span class="o">=</span><span class="n">aa_subseq_df</span><span class="p">,</span>
                                             <span class="n">domain_df</span><span class="o">=</span><span class="n">protein_domain_feature_df</span><span class="p">,</span>
                                             <span class="n">conservation_df</span><span class="o">=</span><span class="n">conservation_features</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="inner_cell">
<div class="input_area">
<div class="highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="n">feature_df</span><span class="p">[</span><span class="n">feature_list</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">feature_list</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
    {% endraw %}

</div>
