---

title: targetdata


keywords: fastai
sidebar: home_sidebar

summary: "Get underlying data to calculate Rule Set 3 target scores"
description: "Get underlying data to calculate Rule Set 3 target scores"
nb_path: "01_targetdata.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 01_targetdata.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">design_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s1">&#39;test_data/sgrna-designs.txt&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Get-amino-acid-sequences">Get amino acid sequences<a class="anchor-link" href="#Get-amino-acid-sequences"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="ensembl_post" class="doc_header"><code>ensembl_post</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetdata.py#L17" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>ensembl_post</code>(<strong><code>ext</code></strong>, <strong><code>data</code></strong>, <strong><code>headers</code></strong>=<em><code>None</code></em>, <strong><code>params</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Generic wrapper for using POST requests to the ensembl rest API</p>
<p>:param ext: str, url extension
:param data: dict, query data
:param headers: dict or None,  meta-information for query
:param params: dict or None, parameters for query
:return: Response object</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="chunks" class="doc_header"><code>chunks</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetdata.py#L38" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>chunks</code>(<strong><code>lst</code></strong>, <strong><code>n</code></strong>)</p>
</blockquote>
<p>Yield successive n-sized chunks from lst.</p>
<p>lst: list
n: int</p>
<p>returns: generator of list chunks</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="post_transcript_sequence_chunk" class="doc_header"><code>post_transcript_sequence_chunk</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetdata.py#L49" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>post_transcript_sequence_chunk</code>(<strong><code>ids</code></strong>, <strong><code>params</code></strong>, <strong><code>headers</code></strong>)</p>
</blockquote>
<p>Helper function for post_transcript_sequence</p>
<p>:param ids: list
:param params: dict
:param headers: dict
:return: dict</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="post_transcript_sequence" class="doc_header"><code>post_transcript_sequence</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetdata.py#L63" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>post_transcript_sequence</code>(<strong><code>ensembl_ids</code></strong>, <strong><code>seq_type</code></strong>=<em><code>'protein'</code></em>, <strong><code>max_queries</code></strong>=<em><code>50</code></em>, <strong><code>n_jobs</code></strong>=<em><code>1</code></em>, <strong>**<code>kwargs</code></strong>)</p>
</blockquote>
<p>Request multiple types of sequence by stable identifier. Supports feature masking and expand options.
Uses <a href="https://rest.ensembl.org/documentation/info/sequence_id_post">https://rest.ensembl.org/documentation/info/sequence_id_post</a></p>
<p>:param ensembl_ids: list of str
:param seq_type: str, one of [genomic, cds, cdna, protein]
:param max_queries: int, maximum number of queries for post
:param n_jobs: int, number of jobs to run in parallel
:param kwargs: additional parameter arguments
:return: list, dict of sequences 5' to 3' in the same orientation as the input transcript</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span><span class="p">(</span><span class="n">post_transcript_sequence</span><span class="p">([</span><span class="s2">&quot;ENSG00000157764&quot;</span><span class="p">,</span> <span class="s2">&quot;ENSG00000248378&quot;</span><span class="p">],</span>
                                <span class="n">seq_type</span><span class="o">=</span><span class="s1">&#39;genomic&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">][</span><span class="s1">&#39;seq&#39;</span><span class="p">][:</span><span class="mi">4</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;CTTC&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 1/1 [00:01&lt;00:00,  1.31s/it]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="build_transcript_aa_seq_df" class="doc_header"><code>build_transcript_aa_seq_df</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetdata.py#L85" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>build_transcript_aa_seq_df</code>(<strong><code>design_df</code></strong>, <strong><code>transcript_id_col</code></strong>=<em><code>'Target Transcript'</code></em>, <strong><code>transcript_len_col</code></strong>=<em><code>'Target Total Length'</code></em>, <strong><code>n_jobs</code></strong>=<em><code>1</code></em>)</p>
</blockquote>
<p>Get amino acid sequence for transcripts of interest</p>
<p>:param design_df: DataFrame
:param transcript_id_col: str, column with ensembl transcript id
:param transcript_len_col: str, column with length of transcript
:param n_jobs: int, number of jobs to use to query transcripts
:return: DataFrame</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">transcript_aa_seq_df</span> <span class="o">=</span> <span class="n">build_transcript_aa_seq_df</span><span class="p">(</span><span class="n">design_df</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">transcript_aa_seq_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting amino acid sequences
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 4/4 [00:00&lt;00:00, 47.11it/s]
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Target Transcript</th>
      <th>Target Total Length</th>
      <th>Transcript Base</th>
      <th>id</th>
      <th>desc</th>
      <th>seq</th>
      <th>molecule</th>
      <th>version</th>
      <th>AA len</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ENST00000259457.8</td>
      <td>834</td>
      <td>ENST00000259457</td>
      <td>ENSP00000259457</td>
      <td>None</td>
      <td>MAAVSVYAPPVGGFSFDNCRRNAVLEADFAKRGYKLPKVRKTGTTI...</td>
      <td>protein</td>
      <td>3</td>
      <td>277</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ENST00000394249.8</td>
      <td>1863</td>
      <td>ENST00000394249</td>
      <td>ENSP00000377793</td>
      <td>None</td>
      <td>MRRSEVLAEESIVCLQKALNHLREIWELIGIPEDQRLQRTEVVKKH...</td>
      <td>protein</td>
      <td>3</td>
      <td>620</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ENST00000361337.3</td>
      <td>2298</td>
      <td>ENST00000361337</td>
      <td>ENSP00000354522</td>
      <td>None</td>
      <td>MSGDHLHNDSQIEADFRLNDSHKHKDKHKDREHRHKEHKKEKDREK...</td>
      <td>protein</td>
      <td>2</td>
      <td>765</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ENST00000368328.5</td>
      <td>267</td>
      <td>ENST00000368328</td>
      <td>ENSP00000357311</td>
      <td>None</td>
      <td>MALSTIVSQRKQIKRKAPRGFLKRVFKRKKPQLRLEKSGDLLVHLN...</td>
      <td>protein</td>
      <td>4</td>
      <td>88</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ENST00000610426.5</td>
      <td>783</td>
      <td>ENST00000610426</td>
      <td>ENSP00000483484</td>
      <td>None</td>
      <td>MPQNEYIELHRKRYGYRLDYHEKKRKKESREAHERSKKAKKMIGLK...</td>
      <td>protein</td>
      <td>1</td>
      <td>260</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>195</th>
      <td>ENST00000339374.11</td>
      <td>1371</td>
      <td>ENST00000339374</td>
      <td>ENSP00000343041</td>
      <td>None</td>
      <td>MAAALKCLLTLGRWCPGLGVAPQARALAALVPGVTQVDNKSGFLQK...</td>
      <td>protein</td>
      <td>6</td>
      <td>456</td>
    </tr>
    <tr>
      <th>196</th>
      <td>ENST00000393047.8</td>
      <td>882</td>
      <td>ENST00000393047</td>
      <td>ENSP00000376767</td>
      <td>None</td>
      <td>MSRIPLGKVLLRNVIRHTDAHNKIQEESDMWKIRELEKQMEDAYRG...</td>
      <td>protein</td>
      <td>3</td>
      <td>293</td>
    </tr>
    <tr>
      <th>197</th>
      <td>ENST00000454402.7</td>
      <td>1023</td>
      <td>ENST00000454402</td>
      <td>ENSP00000408295</td>
      <td>None</td>
      <td>METSALKQQEQPAATKIRNLPWVEKYRPQTLNDLISHQDILSTIQK...</td>
      <td>protein</td>
      <td>2</td>
      <td>340</td>
    </tr>
    <tr>
      <th>198</th>
      <td>ENST00000254998.3</td>
      <td>423</td>
      <td>ENST00000254998</td>
      <td>ENSP00000254998</td>
      <td>None</td>
      <td>MASVDFKTYVDQACRAAEEFVNVYYTTMDKRRRLLSRLYMGTATLV...</td>
      <td>protein</td>
      <td>2</td>
      <td>140</td>
    </tr>
    <tr>
      <th>199</th>
      <td>ENST00000381685.10</td>
      <td>2067</td>
      <td>ENST00000381685</td>
      <td>ENSP00000371101</td>
      <td>None</td>
      <td>MQVSSLNEVKIYSLSCGKSLPEWLSDRKKRALQKKDVDVRRRIELI...</td>
      <td>protein</td>
      <td>5</td>
      <td>688</td>
    </tr>
  </tbody>
</table>
<p>200 rows × 9 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Get-protein-domains">Get protein domains<a class="anchor-link" href="#Get-protein-domains"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="ensembl_get" class="doc_header"><code>ensembl_get</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetdata.py#L121" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>ensembl_get</code>(<strong><code>ext</code></strong>, <strong><code>query</code></strong>=<em><code>None</code></em>, <strong><code>headers</code></strong>=<em><code>None</code></em>, <strong><code>params</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Generic wrapper for using GET requests to the ensembl rest API</p>
<p>ext: str, url extension |
query: str or None, end of url extension specifying species, taxon, esnembl_id etc |
headers: dict or None,  meta-information for query |
params: dict or None, parameters for query |</p>
<p>returns: Response object</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_translation_overlap" class="doc_header"><code>get_translation_overlap</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetdata.py#L143" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_translation_overlap</code>(<strong><code>ensembl_id</code></strong>)</p>
</blockquote>
<p>Get features that overlap with translation, such as protein domains</p>
<p>:param ensembl_id: str
:return: DataFrame</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">brca1_overlap</span> <span class="o">=</span> <span class="n">get_translation_overlap</span><span class="p">(</span><span class="s1">&#39;ENSP00000350283&#39;</span><span class="p">)</span>
<span class="k">assert</span> <span class="s1">&#39;BRCT domain&#39;</span> <span class="ow">in</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">brca1_overlap</span><span class="p">)[</span><span class="s1">&#39;description&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="build_translation_overlap_df" class="doc_header"><code>build_translation_overlap_df</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetdata.py#L156" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>build_translation_overlap_df</code>(<strong><code>protein_ids</code></strong>, <strong><code>n_jobs</code></strong>=<em><code>1</code></em>)</p>
</blockquote>
<p>Get protein domain information</p>
<p>:param protein_ids: list of str, ensemble protein IDs
:param n_jobs: int
:return: DataFrame</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">translation_overlap_df</span> <span class="o">=</span> <span class="n">build_translation_overlap_df</span><span class="p">(</span><span class="n">transcript_aa_seq_df</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">],</span>
                                                      <span class="n">n_jobs</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="n">translation_overlap_df</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting protein domains
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 200/200 [00:35&lt;00:00,  5.61it/s]
</pre>
</div>
</div>

<div class="output_area">


<div class="output_html rendered_html output_subarea output_execute_result">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>feature_type</th>
      <th>hit_end</th>
      <th>type</th>
      <th>cigar_string</th>
      <th>id</th>
      <th>hit_start</th>
      <th>end</th>
      <th>description</th>
      <th>seq_region_name</th>
      <th>hseqname</th>
      <th>interpro</th>
      <th>translation_id</th>
      <th>start</th>
      <th>Transcript Base</th>
      <th>align_type</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>protein_feature</td>
      <td>233</td>
      <td>Gene3D</td>
      <td></td>
      <td>3.60.20.10</td>
      <td>1</td>
      <td>277</td>
      <td>Nucleophile aminohydrolases, N-terminal</td>
      <td>ENSP00000259457</td>
      <td>3.60.20.10</td>
      <td>IPR029055</td>
      <td>976188</td>
      <td>44</td>
      <td>ENST00000259457</td>
      <td>None</td>
    </tr>
    <tr>
      <th>1</th>
      <td>protein_feature</td>
      <td>36</td>
      <td>Pfam</td>
      <td></td>
      <td>PF12465</td>
      <td>1</td>
      <td>271</td>
      <td>Proteasome beta subunit, C-terminal</td>
      <td>ENSP00000259457</td>
      <td>PF12465</td>
      <td>IPR024689</td>
      <td>976188</td>
      <td>235</td>
      <td>ENST00000259457</td>
      <td>None</td>
    </tr>
    <tr>
      <th>2</th>
      <td>protein_feature</td>
      <td>190</td>
      <td>Pfam</td>
      <td></td>
      <td>PF00227</td>
      <td>2</td>
      <td>221</td>
      <td>Proteasome, subunit alpha/beta</td>
      <td>ENSP00000259457</td>
      <td>PF00227</td>
      <td>IPR001353</td>
      <td>976188</td>
      <td>41</td>
      <td>ENST00000259457</td>
      <td>None</td>
    </tr>
    <tr>
      <th>3</th>
      <td>protein_feature</td>
      <td>0</td>
      <td>PRINTS</td>
      <td></td>
      <td>PR00141</td>
      <td>0</td>
      <td>66</td>
      <td>Peptidase T1A, proteasome beta-subunit</td>
      <td>ENSP00000259457</td>
      <td>PR00141</td>
      <td>IPR000243</td>
      <td>976188</td>
      <td>51</td>
      <td>ENST00000259457</td>
      <td>None</td>
    </tr>
    <tr>
      <th>4</th>
      <td>protein_feature</td>
      <td>0</td>
      <td>PRINTS</td>
      <td></td>
      <td>PR00141</td>
      <td>0</td>
      <td>182</td>
      <td>Peptidase T1A, proteasome beta-subunit</td>
      <td>ENSP00000259457</td>
      <td>PR00141</td>
      <td>IPR000243</td>
      <td>976188</td>
      <td>171</td>
      <td>ENST00000259457</td>
      <td>None</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>5275</th>
      <td>protein_feature</td>
      <td>676</td>
      <td>PANTHER</td>
      <td></td>
      <td>PTHR14927</td>
      <td>5</td>
      <td>682</td>
      <td>Nucleolar protein 10/Enp2</td>
      <td>ENSP00000371101</td>
      <td>PTHR14927</td>
      <td>IPR040382</td>
      <td>975385</td>
      <td>1</td>
      <td>ENST00000381685</td>
      <td>None</td>
    </tr>
    <tr>
      <th>5276</th>
      <td>protein_feature</td>
      <td>0</td>
      <td>Seg</td>
      <td></td>
      <td>seg</td>
      <td>0</td>
      <td>474</td>
      <td></td>
      <td>ENSP00000371101</td>
      <td>seg</td>
      <td></td>
      <td>975385</td>
      <td>451</td>
      <td>ENST00000381685</td>
      <td>None</td>
    </tr>
    <tr>
      <th>5277</th>
      <td>protein_feature</td>
      <td>0</td>
      <td>Seg</td>
      <td></td>
      <td>seg</td>
      <td>0</td>
      <td>537</td>
      <td></td>
      <td>ENSP00000371101</td>
      <td>seg</td>
      <td></td>
      <td>975385</td>
      <td>515</td>
      <td>ENST00000381685</td>
      <td>None</td>
    </tr>
    <tr>
      <th>5278</th>
      <td>protein_feature</td>
      <td>0</td>
      <td>Seg</td>
      <td></td>
      <td>seg</td>
      <td>0</td>
      <td>554</td>
      <td></td>
      <td>ENSP00000371101</td>
      <td>seg</td>
      <td></td>
      <td>975385</td>
      <td>543</td>
      <td>ENST00000381685</td>
      <td>None</td>
    </tr>
    <tr>
      <th>5279</th>
      <td>protein_feature</td>
      <td>0</td>
      <td>Seg</td>
      <td></td>
      <td>seg</td>
      <td>0</td>
      <td>661</td>
      <td></td>
      <td>ENSP00000371101</td>
      <td>seg</td>
      <td></td>
      <td>975385</td>
      <td>651</td>
      <td>ENST00000381685</td>
      <td>None</td>
    </tr>
  </tbody>
</table>
<p>5280 rows × 15 columns</p>
</div>
</div>

</div>

</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Store-data-using-parquet-for-later-use">Store data using parquet for later use<a class="anchor-link" href="#Store-data-using-parquet-for-later-use"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="write_transcript_data" class="doc_header"><code>write_transcript_data</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetdata.py#L172" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>write_transcript_data</code>(<strong><code>design_df</code></strong>, <strong><code>transcript_id_col</code></strong>=<em><code>'Target Transcript'</code></em>, <strong><code>transcript_len_col</code></strong>=<em><code>'Target Total Length'</code></em>, <strong><code>n_jobs</code></strong>=<em><code>1</code></em>, <strong><code>overwrite</code></strong>=<em><code>True</code></em>, <strong><code>filepath</code></strong>=<em><code>'./data/target_data/'</code></em>, <strong><code>aa_seq_name</code></strong>=<em><code>'aa_seqs.pq'</code></em>, <strong><code>protein_domain_name</code></strong>=<em><code>'protein_domains.pq'</code></em>)</p>
</blockquote>
<p>Write amino acid sequences and protein domain information to parquet files</p>
<p>:param design_df: DataFrame
:param transcript_id_col: str
:param transcript_len_col: str
:param n_jobs: int
:param overwrite: bool, whether to overwrite existing file
:param filepath: str, directory for output sequences
:param aa_seq_name: str, name of amino acid sequence file
:param protein_domain_name: str, name of protein domain file</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">write_transcript_data</span><span class="p">(</span><span class="n">design_df</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="s1">&#39;./data/target_data/&#39;</span> <span class="o">+</span> <span class="s1">&#39;aa_seqs.pq&#39;</span><span class="p">)</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting amino acid sequences
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 4/4 [00:00&lt;00:00, 2288.22it/s]
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting protein domains
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 200/200 [00:32&lt;00:00,  6.11it/s]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

</div>
 

