---

title: targetfeat


keywords: fastai
sidebar: home_sidebar

summary: "Module to generate target site features"
description: "Module to generate target site features"
nb_path: "02_targetfeat.ipynb"
---
<!--

#################################################
### THIS FILE WAS AUTOGENERATED! DO NOT EDIT! ###
#################################################
# file to edit: 02_targetfeat.ipynb
# command to build the docs after a change: nbdev_build_docs

-->

<div class="container" id="notebook-container">
        
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="kn">from</span> <span class="nn">rs3</span> <span class="kn">import</span> <span class="n">targetdata</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="add_target_columns" class="doc_header"><code>add_target_columns</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L13" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>add_target_columns</code>(<strong><code>design_df</code></strong>, <strong><code>transcript_id_col</code></strong>=<em><code>'Target Transcript'</code></em>, <strong><code>cut_pos_col</code></strong>=<em><code>'Target Cut Length'</code></em>)</p>
</blockquote>
<p>Add ['AA Index' and 'Transcript Base'] to design df</p>
<p>:param design_df: DataFrame
:return: DataFrame</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">design_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_table</span><span class="p">(</span><span class="s1">&#39;test_data/sgrna-designs.txt&#39;</span><span class="p">)</span>
<span class="n">design_targ_df</span> <span class="o">=</span> <span class="n">add_target_columns</span><span class="p">(</span><span class="n">design_df</span><span class="p">)</span>
<span class="k">assert</span> <span class="s1">&#39;AA Index&#39;</span> <span class="ow">in</span> <span class="n">design_targ_df</span><span class="o">.</span><span class="n">columns</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Position-Features">Position Features<a class="anchor-link" href="#Position-Features"> </a></h2><p>The first feature class we consider is where the guide targets within the annotated transcript</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_position_features" class="doc_header"><code>get_position_features</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L26" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_position_features</code>(<strong><code>sg_df</code></strong>, <strong><code>id_cols</code></strong>)</p>
</blockquote>
<p>Get  features ['Target Cut %', 'sense']</p>
<p>:param sg_df: DataFrame
:param id_cols: list
:return: DataFrame</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Amino-Acid-Features">Amino Acid Features<a class="anchor-link" href="#Amino-Acid-Features"> </a></h2><p>We calculate a set of features from the amino acid sequence around the cutsite itself</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">aas</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;D&#39;</span><span class="p">,</span> <span class="s1">&#39;E&#39;</span><span class="p">,</span> <span class="s1">&#39;F&#39;</span><span class="p">,</span>
       <span class="s1">&#39;G&#39;</span><span class="p">,</span> <span class="s1">&#39;H&#39;</span><span class="p">,</span> <span class="s1">&#39;I&#39;</span><span class="p">,</span> <span class="s1">&#39;K&#39;</span><span class="p">,</span> <span class="s1">&#39;L&#39;</span><span class="p">,</span>
       <span class="s1">&#39;M&#39;</span><span class="p">,</span> <span class="s1">&#39;N&#39;</span><span class="p">,</span> <span class="s1">&#39;P&#39;</span><span class="p">,</span> <span class="s1">&#39;Q&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">,</span>
       <span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;V&#39;</span><span class="p">,</span> <span class="s1">&#39;W&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;*&#39;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_one_aa_frac" class="doc_header"><code>get_one_aa_frac</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L38" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_one_aa_frac</code>(<strong><code>feature_dict</code></strong>, <strong><code>aa_sequence</code></strong>, <strong><code>aas</code></strong>)</p>
</blockquote>
<p>Get fraction of single aa</p>
<p>:param feature_dict: dict, feature dictionary
:param aa_sequence: str, amino acid sequence
:param aas: list, list of amino acids</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">one_aa_ft</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">get_one_aa_frac</span><span class="p">(</span><span class="n">one_aa_ft</span><span class="p">,</span> <span class="s1">&#39;ACDG*-&#39;</span><span class="p">,</span> <span class="n">aas</span><span class="p">)</span>
<span class="k">assert</span> <span class="n">one_aa_ft</span><span class="p">[</span><span class="s1">&#39;A&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">1</span><span class="o">/</span><span class="mi">6</span>
<span class="k">assert</span> <span class="n">one_aa_ft</span><span class="p">[</span><span class="s1">&#39;Q&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="mi">0</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_aa_aromaticity" class="doc_header"><code>get_aa_aromaticity</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L50" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_aa_aromaticity</code>(<strong><code>feature_dict</code></strong>, <strong><code>analyzed_seq</code></strong>)</p>
</blockquote>
<p>Get fraction of aromatic amino acids in a sequence.</p>
<p>Phe (F) + Trp (W) + Tyr (Y)</p>
<p>:param feature_dict:
:param analyzed_seq: ProteinAnalysis object</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_aa_hydrophobicity" class="doc_header"><code>get_aa_hydrophobicity</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L61" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_aa_hydrophobicity</code>(<strong><code>feature_dict</code></strong>, <strong><code>analyzed_seq</code></strong>)</p>
</blockquote>
<p>Grand Average of Hydropathy</p>
<p>The GRAVY value is calculated by adding the hydropathy value for each residue and dividing
 by the length of the sequence (Kyte and Doolittle; 1982). The larger the number, the more hydrophobic the
 amino acid</p>
<p>:param feature_dict: dict
:param analyzed_seq: ProteinAnalysis object</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_aa_ip" class="doc_header"><code>get_aa_ip</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L74" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_aa_ip</code>(<strong><code>feature_dict</code></strong>, <strong><code>analyzed_seq</code></strong>)</p>
</blockquote>
<p>Get the Isoelectric Point of an amino acid sequence</p>
<p>Charge of amino acid</p>
<p>:param feature_dict: dict
:param analyzed_seq: ProteinAnalysis object</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_aa_secondary_structure" class="doc_header"><code>get_aa_secondary_structure</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L85" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_aa_secondary_structure</code>(<strong><code>feature_dict</code></strong>, <strong><code>analyzed_seq</code></strong>)</p>
</blockquote>
<p>Get the fraction of amion acids that tend to be in a helix, turn or sheet</p>
<p>:param feature_dict: dict
:param analyzed_seq: ProteinAnalysis object</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">aa_biochemical_fts1</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">get_aa_aromaticity</span><span class="p">(</span><span class="n">aa_biochemical_fts1</span><span class="p">,</span> <span class="n">ProteinAnalysis</span><span class="p">(</span><span class="s1">&#39;FWYA&#39;</span><span class="p">))</span>
<span class="n">aa_biochemical_fts2</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">get_aa_aromaticity</span><span class="p">(</span><span class="n">aa_biochemical_fts2</span><span class="p">,</span> <span class="n">ProteinAnalysis</span><span class="p">(</span><span class="s1">&#39;AAAA&#39;</span><span class="p">))</span>
<span class="k">assert</span> <span class="n">aa_biochemical_fts1</span><span class="p">[</span><span class="s1">&#39;Aromaticity&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">aa_biochemical_fts2</span><span class="p">[</span><span class="s1">&#39;Aromaticity&#39;</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="featurize_aa_seqs" class="doc_header"><code>featurize_aa_seqs</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L95" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>featurize_aa_seqs</code>(<strong><code>aa_sequences</code></strong>, <strong><code>features</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Get feature DataFrame for a list of amino acid sequences</p>
<p>:param aa_sequences: list of str
:param features: list or None
:return: DataFrame</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">ft_dict_df</span> <span class="o">=</span> <span class="n">featurize_aa_seqs</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">([</span><span class="s1">&#39;ACDG*-&#39;</span><span class="p">,</span> <span class="s1">&#39;CDG*--&#39;</span><span class="p">,</span> <span class="s1">&#39;LLLLLL&#39;</span><span class="p">]))</span>
<span class="k">assert</span> <span class="n">ft_dict_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="s1">&#39;LLLLLL&#39;</span><span class="p">,</span> <span class="s1">&#39;Hydrophobicity&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">ft_dict_df</span><span class="p">[</span><span class="s1">&#39;Hydrophobicity&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_amino_acid_features" class="doc_header"><code>get_amino_acid_features</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L130" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_amino_acid_features</code>(<strong><code>sg_designs</code></strong>, <strong><code>aa_seq_df</code></strong>, <strong><code>width</code></strong>, <strong><code>features</code></strong>, <strong><code>id_cols</code></strong>, <strong><code>transcript_base_col</code></strong>=<em><code>'Transcript Base'</code></em>, <strong><code>target_transcript_col</code></strong>=<em><code>'Target Transcript'</code></em>, <strong><code>aa_index_col</code></strong>=<em><code>'AA Index'</code></em>)</p>
</blockquote>
<p>Featurize amino acid sequences</p>
<p>:param sg_designs: DataFrame
:param aa_seq_df: DataFrame, Transcript Base and (AA) seq
:param width: int, length on each side of the cut site
:param feature: list
:param transcript_base_col: str
:param target_transcript_col: str
:param aa_index_col: str
:return: DataFrame</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">aa_seq_df</span> <span class="o">=</span> <span class="n">targetdata</span><span class="o">.</span><span class="n">build_transcript_aa_seq_df</span><span class="p">(</span><span class="n">design_targ_df</span><span class="p">,</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">get_rev_comp</span><span class="p">(</span><span class="n">sgrna</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Get reverse compliment of a guide&quot;&quot;&quot;</span>
    <span class="n">nt_map</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;A&#39;</span><span class="p">:</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">:</span> <span class="s1">&#39;A&#39;</span><span class="p">,</span> <span class="s1">&#39;G&#39;</span><span class="p">:</span> <span class="s1">&#39;C&#39;</span><span class="p">,</span> <span class="s1">&#39;C&#39;</span><span class="p">:</span> <span class="s1">&#39;G&#39;</span><span class="p">}</span>
    <span class="n">rev_comp</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
    <span class="k">for</span> <span class="n">nt</span> <span class="ow">in</span> <span class="n">sgrna</span><span class="p">:</span>
        <span class="n">rev_comp</span> <span class="o">+=</span> <span class="n">nt_map</span><span class="p">[</span><span class="n">nt</span><span class="p">]</span>
    <span class="n">rev_comp</span> <span class="o">=</span> <span class="n">rev_comp</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">rev_comp</span>

<span class="n">aa_features</span> <span class="o">=</span> <span class="n">get_amino_acid_features</span><span class="p">(</span><span class="n">design_targ_df</span><span class="p">,</span> <span class="n">aa_seq_df</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
                                      <span class="n">features</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Pos. Ind. 1mer&#39;</span><span class="p">,</span> <span class="s1">&#39;Pos. Ind. 2mer&#39;</span><span class="p">,</span> <span class="s1">&#39;Pos. Dep. 1mer&#39;</span><span class="p">,</span>
                                                <span class="s1">&#39;Hydrophobicity&#39;</span><span class="p">,</span> <span class="s1">&#39;Aromaticity&#39;</span><span class="p">,</span>
                                                <span class="s1">&#39;Isoelectric Point&#39;</span><span class="p">,</span> <span class="s1">&#39;Secondary Structure&#39;</span><span class="p">],</span>
                                      <span class="n">id_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sgRNA Context Sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Target Cut Length&#39;</span><span class="p">,</span>
                                               <span class="s1">&#39;Target Transcript&#39;</span><span class="p">,</span> <span class="s1">&#39;Orientation&#39;</span><span class="p">])</span>
<span class="k">assert</span> <span class="p">(</span><span class="n">aa_features</span><span class="p">[</span><span class="s1">&#39;AA Subsequence&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">len</span><span class="p">()</span> <span class="o">==</span> <span class="mi">20</span><span class="p">)</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting amino acid sequences
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 4/4 [00:02&lt;00:00,  1.79it/s]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">codon_map_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;test_data/codon_map.csv&#39;</span><span class="p">)</span>
<span class="n">codon_map</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">codon_map_df</span><span class="p">[</span><span class="s1">&#39;Amino Acid&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">codon_map_df</span><span class="p">[</span><span class="s1">&#39;Codon&#39;</span><span class="p">])</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
<span class="n">row</span> <span class="o">=</span> <span class="n">aa_features</span><span class="o">.</span><span class="n">sample</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:]</span>
<span class="n">subseq</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;AA Subsequence&#39;</span><span class="p">]</span>
<span class="n">context</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;sgRNA Context Sequence&#39;</span><span class="p">]</span>
<span class="n">rc_context</span> <span class="o">=</span> <span class="n">get_rev_comp</span><span class="p">(</span><span class="n">context</span><span class="p">)</span>
<span class="n">translations</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="n">rc_translations</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>
<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">]:</span>
    <span class="n">translations</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">codon_map</span><span class="p">[</span><span class="n">context</span><span class="p">[</span><span class="n">j</span><span class="p">:</span><span class="n">j</span><span class="o">+</span><span class="mi">3</span><span class="p">]]</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">context</span><span class="p">),</span> <span class="mi">3</span><span class="p">)</span>
                               <span class="k">if</span> <span class="p">(</span><span class="n">j</span> <span class="o">+</span> <span class="mi">3</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">context</span><span class="p">)])</span>
    <span class="n">rc_translations</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">codon_map</span><span class="p">[</span><span class="n">rc_context</span><span class="p">[</span><span class="n">j</span><span class="p">:</span><span class="n">j</span><span class="o">+</span><span class="mi">3</span><span class="p">]]</span> <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">rc_context</span><span class="p">),</span> <span class="mi">3</span><span class="p">)</span>
                                  <span class="k">if</span> <span class="p">(</span><span class="n">j</span> <span class="o">+</span> <span class="mi">3</span><span class="p">)</span> <span class="o">&lt;=</span> <span class="nb">len</span><span class="p">(</span><span class="n">rc_context</span><span class="p">)])</span>
<span class="k">assert</span> <span class="p">((</span><span class="n">translations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">translations</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">translations</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">)</span> <span class="ow">or</span>
        <span class="p">(</span><span class="n">rc_translations</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">rc_translations</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">rc_translations</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="ow">in</span> <span class="n">subseq</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Protein-Domain-Features">Protein Domain Features<a class="anchor-link" href="#Protein-Domain-Features"> </a></h2>
</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="get_protein_domain_features" class="doc_header"><code>get_protein_domain_features</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L173" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>get_protein_domain_features</code>(<strong><code>sg_design_df</code></strong>, <strong><code>protein_domains</code></strong>, <strong><code>sources</code></strong>, <strong><code>id_cols</code></strong>, <strong><code>transcript_base_col</code></strong>=<em><code>'Transcript Base'</code></em>, <strong><code>aa_index_col</code></strong>=<em><code>'AA Index'</code></em>, <strong><code>domain_type_col</code></strong>=<em><code>'type'</code></em>, <strong><code>domain_start_col</code></strong>=<em><code>'start'</code></em>, <strong><code>domain_end_col</code></strong>=<em><code>'end'</code></em>)</p>
</blockquote>
<p>Get binary dataframe of protein domains</p>
<p>:param sg_design_df: DataFrame, with columns [transcript_base_col, aa_index_col]
:param protein_domains: DataFrame, with columns [transcript_base_col, domain_type_col]
:param sources: list. list of database types to include
:param id_cols: list
:param transcript_base_col: str
:param aa_index_col: str
:param domain_type_col: str
:param domain_start_col: str
:param domain_end_col: str
:return: DataFrame, with binary features for protein domains</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">domain_df</span> <span class="o">=</span> <span class="n">targetdata</span><span class="o">.</span><span class="n">build_translation_overlap_df</span><span class="p">(</span><span class="n">aa_seq_df</span><span class="p">[</span><span class="s1">&#39;id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span> <span class="n">n_jobs</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
<span class="n">protein_domain_feature_df</span> <span class="o">=</span> <span class="n">get_protein_domain_features</span><span class="p">(</span><span class="n">design_targ_df</span><span class="p">,</span> <span class="n">domain_df</span><span class="p">,</span> <span class="n">sources</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                                                        <span class="n">id_cols</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sgRNA Context Sequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Target Cut Length&#39;</span><span class="p">,</span>
                                                                 <span class="s1">&#39;AA Index&#39;</span><span class="p">,</span> <span class="s1">&#39;Target Transcript&#39;</span><span class="p">,</span> <span class="s1">&#39;Orientation&#39;</span><span class="p">])</span>
</pre></div>

    </div>
</div>
</div>

<div class="output_wrapper">
<div class="output">

<div class="output_area">

<div class="output_subarea output_stream output_stdout output_text">
<pre>Getting protein domains
</pre>
</div>
</div>

<div class="output_area">

<div class="output_subarea output_stream output_stderr output_text">
<pre>100%|██████████| 200/200 [00:52&lt;00:00,  3.79it/s]
</pre>
</div>
</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="k">assert</span> <span class="n">protein_domain_feature_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">protein_domain_feature_df</span><span class="p">[</span><span class="s1">&#39;sgRNA Context Sequence&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;AAAAGAGCCATGAATCTAAACATCAGGAAT&#39;</span><span class="p">,</span>
                                     <span class="p">[</span><span class="s1">&#39;PANTHER&#39;</span><span class="p">,</span> <span class="s1">&#39;ncoils&#39;</span><span class="p">,</span> <span class="s1">&#39;Seg&#39;</span><span class="p">,</span> <span class="s1">&#39;MobiDBLite&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="mi">4</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Combining-target-features">Combining target features<a class="anchor-link" href="#Combining-target-features"> </a></h2><p>We'll combine, the position, amino acid and domain feature matrices into a single target feature matrix</p>

</div>
</div>
</div>
    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

<div class="output_wrapper">
<div class="output">

<div class="output_area">


<div class="output_markdown rendered_html output_subarea ">
<h4 id="build_target_feature_df" class="doc_header"><code>build_target_feature_df</code><a href="https://github.com/gpp-rnd/rs3/tree/master/rs3/targetfeat.py#L217" class="source_link" style="float:right">[source]</a></h4><blockquote><p><code>build_target_feature_df</code>(<strong><code>sg_designs</code></strong>, <strong><code>features</code></strong>=<em><code>None</code></em>, <strong><code>aa_seq_df</code></strong>=<em><code>None</code></em>, <strong><code>aa_width</code></strong>=<em><code>16</code></em>, <strong><code>aa_features</code></strong>=<em><code>None</code></em>, <strong><code>protein_domain_df</code></strong>=<em><code>None</code></em>, <strong><code>protein_domain_sources</code></strong>=<em><code>None</code></em>, <strong><code>id_cols</code></strong>=<em><code>None</code></em>)</p>
</blockquote>
<p>Build the feature matrix for the sgRNA target site</p>
<p>:param sg_designs: DataFrame
:param features: list
:param aa_seq_df: DataFrame
:param aa_width: int
:param aa_features: list
:param protein_domain_df: DataFrame
:param protein_domain_sources: list or None. Defaults to all sources except Sifts
:return: (feature_df, feature_list)
    feature_df: DataFrame
    feature_list: list</p>

</div>

</div>

</div>
</div>

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">

</div>
    {% endraw %}

    {% raw %}
    
<div class="cell border-box-sizing code_cell rendered">
<div class="input">

<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-ipython3"><pre><span></span><span class="n">feature_df</span><span class="p">,</span> <span class="n">feature_list</span> <span class="o">=</span> <span class="n">build_target_feature_df</span><span class="p">(</span><span class="n">sg_designs</span><span class="o">=</span><span class="n">design_df</span><span class="p">,</span>
                                                   <span class="n">aa_seq_df</span><span class="o">=</span><span class="n">aa_seq_df</span><span class="p">,</span>
                                                   <span class="n">protein_domain_df</span><span class="o">=</span><span class="n">domain_df</span><span class="p">)</span>
<span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="n">feature_list</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">20</span>
<span class="k">assert</span> <span class="n">feature_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">design_df</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>

    </div>
</div>
</div>

</div>
    {% endraw %}

</div>
 

